domain/modelconfig/handlers/secrethandler.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package handlers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+
+	coremodel "github.com/juju/juju/core/model"
+	"github.com/juju/juju/domain/modeldefaults"
+	backenderrors "github.com/juju/juju/domain/secretbackend/errors"
+	"github.com/juju/juju/environs/config"
+)
+
+// SecretBackendState provides access to the state where secret backends are stored.
+type SecretBackendState interface {
+	SetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID, backendName string) error
+	GetModelSecretBackendName(ctx context.Context, modelUUID coremodel.UUID) (string, error)
+}
+
+// SecretBackendHandler implements ConfigHandler.
+type SecretBackendHandler struct {
+	Defaults     modeldefaults.Defaults
+	BackendState SecretBackendState
+	ModelUUID    coremodel.UUID
+}
+
+// Name is the handler name.
+func (h SecretBackendHandler) Name() string {
+	return "model secret backend"
+}
+
+// OnSave is run when saving model config; any secret
+// backend is removed from rawCfg and used to update the
+// model secret backend config table.
+func (h SecretBackendHandler) OnSave(ctx context.Context, rawCfg map[string]any, removeAttrs []string) error {
+	// If removing, set to the model default value.
+	toRemove := set.NewStrings(removeAttrs...)
+	if toRemove.Contains(config.SecretBackendKey) {
```

Comment:
I don't understand this. This is violating the services layer for model config. The service layer for model config should build up a completely new view of what the model config will look like including the new defaults of the removed attributes. Only then should this handler be running.

I think the func signature is a little off in this design as you should only be operating on what was and what will be. The service layer handles view building.

The handler then just becomes has the value changed between the two revisions and if yes run the checks and persist the state.

Reply:
We're not on the same page - I suspect there's a slight misunderstanding of what's going on here. Easier to discuss f2f. The service layer is indeed building up a new view of what goes into the final config, but it is doing so as part of a workflow to also maintain the relational data model to reflect the intent of the those attributes which are presented as config as far as the UX is concerned, but which really are part of the relational model.

Reply:
This is now removed. The handler does not use the defaults anymore.